### ISSUE TITLE
Reproduce

### ISSUE THREAD

**Issue by Linda6521** [boschresearch/MultiADS#21]
When I trained on VisA to get the weight and then used it  to test the MVTec-AD dataset , i reproduced a very low result, could you please tell me the reason. Thanks.

**Comment by hk-zh** [comment]
Could I kindly ask which epoch's checkpoint you used for the test?

**Comment by Linda6521** [comment]
I trained on VisA dataset for 10 epoch, and then used both the 1st epoch's checkpoint and 10th epoch's checkpoint to test on the MVTec-AD respectively, but the results are all very low.
26-02-18 02:18:38.865 - INFO: checkpoint_path: /root/MultiADS/visackpt/epoch_1.pth
26-02-18 02:18:38.865 - INFO: config_path: ./open_clip/model_configs/ViT-L-14-336.json
26-02-18 02:18:38.865 - INFO: dataset: mvtec
26-02-18 02:18:38.865 - INFO: model: ViT-L-14-336
26-02-18 02:18:38.865 - INFO: pretrained: openai
26-02-18 02:18:38.865 - INFO: features_list: [6, 12, 18, 24]
26-02-18 02:18:38.865 - INFO: image_size: 518
26-02-18 02:18:38.865 - INFO: mode: zero_shot
26-02-18 02:18:38.865 - INFO: quantile: 0.0001
26-02-18 02:18:38.865 - INFO: seed: 42
26-02-18 03:04:54.600 - INFO: 
| objects    |   auroc_px |   f1_px |   ap_px |   aupro |   auroc_sp |   f1_sp |   ap_sp |
|:-----------|-----------:|--------:|--------:|--------:|-----------:|--------:|--------:|
| bottle     |       23.9 |    10.9 |     3.5 |     0.3 |       60.4 |    89.7 |    82.7 |
| cable      |       36.1 |     5.6 |     2   |     1.4 |       43.1 |    76   |    56.2 |
| capsule    |       16.9 |     1.8 |     0.5 |     0.1 |       35.8 |    90.5 |    77.5 |
| carpet     |        8.4 |     3.2 |     0.8 |     0.5 |       49.5 |    86.4 |    76.4 |
| grid       |       17.7 |     1.4 |     0.4 |     0.3 |       48.2 |    84.4 |    76.1 |
| hazelnut   |       14.5 |     4.2 |     1.2 |     0   |       68.7 |    82.1 |    78.3 |
| leather    |        6.2 |     1.3 |     0.3 |     0.2 |       58   |    85.2 |    81.9 |
| metal_nut  |       39   |    21.3 |     8.7 |     0.9 |       43.5 |    89.9 |    77   |
| pill       |       29   |     6.5 |     2.1 |     0.1 |       49   |    91.6 |    85.4 |
| screw      |       11.3 |     0.5 |     0.1 |     0   |       50.2 |    85.3 |    75.3 |
| tile       |       18.6 |    13.2 |     4   |     1   |       65.2 |    83.6 |    81.2 |
| toothbrush |       20.3 |     3   |     0.9 |     0   |       60.8 |    84.5 |    80.8 |
| transistor |       37.1 |     9.2 |     3.5 |     3.7 |       57.4 |    59.6 |    50.9 |
| wood       |       13.5 |     7.4 |     2.1 |     0.7 |       67.6 |    86.3 |    88.7 |
| zipper     |       27.1 |     4   |     1.3 |     0.4 |       53.2 |    88.5 |    83.6 |
| mean       |       21.3 |     6.2 |     2.1 |     0.6 |       54   |    84.2 |    76.8 |and 
26-02-18 05:56:51.120 - INFO: checkpoint_path: /root/MultiADS/visackpt/epoch_10.pth
26-02-18 05:56:51.120 - INFO: config_path: ./open_clip/model_configs/ViT-L-14-336.json
26-02-18 05:56:51.121 - INFO: dataset: mvtec
26-02-18 05:56:51.121 - INFO: model: ViT-L-14-336
26-02-18 05:56:51.121 - INFO: pretrained: openai
26-02-18 05:56:51.121 - INFO: features_list: [6, 12, 18, 24]
26-02-18 05:56:51.121 - INFO: image_size: 518
26-02-18 05:56:51.121 - INFO: mode: zero_shot
26-02-18 05:56:51.121 - INFO: quantile: 0.0001
26-02-18 05:56:51.121 - INFO: seed: 42
26-02-18 06:44:24.346 - INFO: 
| objects    |   auroc_px |   f1_px |   ap_px |   aupro |   auroc_sp |   f1_sp |   ap_sp |
|:-----------|-----------:|--------:|--------:|--------:|-----------:|--------:|--------:|
| bottle     |       21.5 |    10.9 |     3.4 |     0.4 |       54.4 |    86.3 |    80   |
| cable      |       33.6 |     5.6 |     1.9 |     1.2 |       62   |    78   |    69.5 |
| capsule    |       16.8 |     1.8 |     0.5 |     0.1 |       52.7 |    90.5 |    84.5 |
| carpet     |        5.4 |     3.2 |     1.8 |     4.1 |       75.2 |    88   |    91.5 |
| grid       |       13.7 |     1.4 |     0.4 |     0.4 |       37.8 |    84.4 |    64.5 |
| hazelnut   |        8.7 |     4.2 |     1.2 |     0.3 |       83.5 |    86.7 |    89.3 |
| leather    |        3.2 |     1.3 |     0.6 |     0.8 |       78.9 |    85.8 |    92.4 |
| metal_nut  |       36   |    21.1 |     8.8 |     2.8 |       70.1 |    89.4 |    91.1 |
| pill       |       25.5 |     6.5 |     2.1 |     0.1 |       45.9 |    91.6 |    85.6 |
| screw      |        7.8 |     0.5 |     0.1 |     0.1 |       31.3 |    85.6 |    68.1 |
| tile       |       18.4 |    13.2 |     4.2 |     1.2 |       71.3 |    84   |    88.2 |
| toothbrush |       18.7 |     3   |     1.3 |     1.6 |       52.2 |    83.3 |    78.6 |
| transistor |       34.1 |     9.2 |     3.3 |     3.4 |       63.8 |    57.8 |    61.7 |
| wood       |        8.7 |     7.4 |     2   |     0.2 |       53.9 |    87   |    82.3 |
| zipper     |       20.8 |     4   |     1.2 |     0.5 |       55.9 |    88.1 |    83.9 |
| mean       |       18.2 |     6.2 |     2.2 |     1.1 |       59.3 |    84.4 |    80.7 |
I used the default configuration in the github repository, so I wonder why.

**Comment by hk-zh** [comment]
seems that something is wrong with your pixel-level results. Shouldn't be so low. Could you double-check the testing settings and the visa data you used?

**Comment by Linda6521** [comment]
I used the VisA_20220922 dataset and used the meta.json in the project. 
parser.add_argument("--mode", type=str, default="zero_shot", help="zero shot or few shot or domain adaption")
    # few shot
    parser.add_argument("--k_shot", type=int, default=10, help="e.g., 10-shot, 5-shot, 1-shot") I tested the zero shot mode and tried the visa ckpt on test.py again but got  the same result

**Comment by Linda6521** [comment]
Or could you kindly provide the results when trained on VisA and tested on MVTec-AD?
Because when trained on the  MVTec-AD and tested on the other datasets, my reproduced results are similar to your paper's, but when trained on VisA and tested on MVTec-AD, it had a huge discrepancy.

**Comment by hk-zh** [comment]
Hi, we retrained and tested MultiADS by 
```
python test.py --dataset mvtec --data_path ./data/mvtec --save_path ./results/mvtec/zero-shot/ --checkpoint_path ./exps/visa_default/epoch_1
python test.py --dataset mvtec --data_path ./data/mvtec --save_path ./results/mvtec/zero-shot/ --checkpoint_path ./exps/visa_default/epoch_2
python test.py --dataset mvtec --data_path ./data/mvtec --save_path ./results/mvtec/zero-shot/ --checkpoint_path ./exps/visa_default/epoch_3
```
These were the results we got:

26-02-20 23:49:23.683 - INFO: data_path: ./data/mvtec
26-02-20 23:49:23.698 - INFO: save_path: ./results/mvtec/zero_shot/
26-02-20 23:49:23.698 - INFO: checkpoint_path: ./exps/visa_default/epoch_1.pth
26-02-20 23:49:23.698 - INFO: config_path: ./open_clip/model_configs/ViT-L-14-336.json
26-02-20 23:49:23.698 - INFO: dataset: mvtec
26-02-20 23:49:23.698 - INFO: model: ViT-L-14-336
26-02-20 23:49:23.698 - INFO: pretrained: openai
26-02-20 23:49:23.698 - INFO: features_list: [6, 12, 18, 24]
26-02-20 23:49:23.698 - INFO: image_size: 518
26-02-20 23:49:23.698 - INFO: mode: zero_shot
26-02-20 23:49:23.698 - INFO: quantile: 0.0001
26-02-20 23:49:23.699 - INFO: seed: 42
26-02-21 00:22:07.200 - INFO: 

| objects    |   auroc_px |   f1_px |   ap_px |   aupro |   auroc_sp |   f1_sp |   ap_sp |
|:-----------|-----------:|--------:|--------:|--------:|-----------:|--------:|--------:|
| bottle     |       72.7 |    39   |    36.4 |    58.1 |       90.1 |    90   |    97.1 |
| cable      |       65   |    19.9 |    11.8 |    55   |       72.4 |    77.7 |    83   |
| capsule    |       90.5 |    26.5 |    20.8 |    68.1 |       72   |    90.5 |    93.4 |
| carpet     |       95.4 |    60.3 |    55   |    69.1 |       98.9 |    98.3 |    99.7 |
| grid       |       85.6 |    36.4 |    22.4 |    53.3 |       96   |    94.8 |    98.7 |
| hazelnut   |       96   |    53.3 |    50   |    77.8 |       97   |    93.6 |    98.4 |
| leather    |       97.9 |    44.7 |    42.1 |    75.3 |       99.3 |    99.5 |    99.7 |
| metal_nut  |       82.2 |    39.4 |    33.3 |    65.5 |       68.3 |    89.4 |    92.4 |
| pill       |       79.2 |    23.2 |    20.1 |    77.8 |       83.4 |    91.6 |    96.8 |
| screw      |       97.9 |    34.9 |    29   |    83.7 |       75.6 |    85.6 |    90.5 |
| tile       |       74.7 |    33.8 |    30.9 |    56.6 |       94   |    94.8 |    96.9 |
| toothbrush |       91.2 |    38.7 |    26.5 |    79.7 |       92.2 |    93.5 |    96.9 |
| transistor |       57.9 |    15.8 |     8.8 |    47.8 |       66.8 |    63.6 |    65.3 |
| wood       |       90.4 |    52.5 |    48.9 |    59.3 |       95.6 |    93.9 |    98.7 |
| zipper     |       86.7 |    24.9 |    21.3 |    39.7 |       76   |    88.1 |    93.4 |
| mean       |       84.2 |    36.2 |    30.5 |    64.4 |       85.2 |    89.7 |    93.4 |

26-02-21 00:22:33.447 - INFO: data_path: ./data/mvtec
26-02-21 00:22:33.447 - INFO: save_path: ./results/mvtec/zero_shot/
26-02-21 00:22:33.447 - INFO: checkpoint_path: ./exps/visa_default/epoch_2.pth
26-02-21 00:22:33.447 - INFO: config_path: ./open_clip/model_configs/ViT-L-14-336.json
26-02-21 00:22:33.447 - INFO: dataset: mvtec
26-02-21 00:22:33.447 - INFO: model: ViT-L-14-336
26-02-21 00:22:33.447 - INFO: pretrained: openai
26-02-21 00:22:33.447 - INFO: features_list: [6, 12, 18, 24]
26-02-21 00:22:33.448 - INFO: image_size: 518
26-02-21 00:22:33.448 - INFO: mode: zero_shot
26-02-21 00:22:33.448 - INFO: quantile: 0.0001
26-02-21 00:22:33.448 - INFO: seed: 42

| objects    |   auroc_px |   f1_px |   ap_px |   aupro |   auroc_sp |   f1_sp |   ap_sp |
|:-----------|-----------:|--------:|--------:|--------:|-----------:|--------:|--------:|
| bottle     |       70.3 |    44.9 |    39.4 |    54.5 |       88.9 |    90.8 |    96.8 |
| cable      |       70.5 |    21.8 |    13.6 |    39.6 |       68.3 |    76   |    81.4 |
| capsule    |       95.7 |    33.6 |    30.1 |    52.5 |       84.2 |    91.6 |    96.6 |
| carpet     |       97.4 |    63.8 |    59.5 |    55.6 |       99.3 |    98.9 |    99.8 |
| grid       |       95.1 |    42.7 |    33.8 |    33.3 |       97.2 |    96.6 |    99.1 |
| hazelnut   |       96.5 |    50.8 |    48.3 |    72.8 |       95.5 |    92.5 |    97.5 |
| leather    |       99.1 |    42.9 |    44.9 |    74.5 |       99.4 |    99.5 |    99.8 |
| metal_nut  |       80.4 |    37.2 |    32.3 |    58   |       76.4 |    89.9 |    94.4 |
| pill       |       84.2 |    28.1 |    24.5 |    72.2 |       80.1 |    91.6 |    95.9 |
| screw      |       97.9 |    35.7 |    29.7 |    74.6 |       75.7 |    86.2 |    90.9 |
| tile       |       83.8 |    48.8 |    46.1 |    37.2 |       94.8 |    94.8 |    97.4 |
| toothbrush |       94.7 |    45.1 |    36.2 |    69   |       90.8 |    92.9 |    97   |
| transistor |       63.5 |    19.8 |    11.2 |    28.4 |       75.2 |    68   |    65.6 |
| wood       |       95   |    58.5 |    56.4 |    49.3 |       94.8 |    95.2 |    98.2 |
| zipper     |       82.7 |    17.4 |    15.5 |    19.9 |       82.5 |    88.1 |    95.5 |
| mean       |       87.1 |    39.4 |    34.8 |    52.8 |       86.9 |    90.2 |    93.7 |

26-02-21 00:55:22.333 - INFO: data_path: ./data/mvtec
26-02-21 00:55:22.333 - INFO: save_path: ./results/mvtec/zero_shot/
26-02-21 00:55:22.333 - INFO: checkpoint_path: ./exps/visa_default/epoch_3.pth
26-02-21 00:55:22.333 - INFO: config_path: ./open_clip/model_configs/ViT-L-14-336.json
26-02-21 00:55:22.334 - INFO: dataset: mvtec
26-02-21 00:55:22.334 - INFO: model: ViT-L-14-336
26-02-21 00:55:22.334 - INFO: pretrained: openai
26-02-21 00:55:22.334 - INFO: features_list: [6, 12, 18, 24]
26-02-21 00:55:22.334 - INFO: image_size: 518
26-02-21 00:55:22.334 - INFO: mode: zero_shot
26-02-21 00:55:22.334 - INFO: quantile: 0.0001
26-02-21 00:55:22.334 - INFO: seed: 42
26-02-21 01:27:38.879 - INFO: 

| objects    |   auroc_px |   f1_px |   ap_px |   aupro |   auroc_sp |   f1_sp |   ap_sp |
|:-----------|-----------:|--------:|--------:|--------:|-----------:|--------:|--------:|
| bottle     |       69.4 |    37.7 |    35.6 |    51.7 |       86.7 |    87.7 |    96.1 |
| cable      |       66.9 |    24.9 |    15.6 |    42.2 |       67   |    76   |    80   |
| capsule    |       94.6 |    31.1 |    27.4 |    57.6 |       83   |    91.1 |    96.2 |
| carpet     |       97.6 |    64   |    59.5 |    48.5 |       99.3 |    97.8 |    99.8 |
| grid       |       94.7 |    41.3 |    32.3 |    31.6 |       97.7 |    95.6 |    99.2 |
| hazelnut   |       96.8 |    54.7 |    54.9 |    82   |       96.9 |    93.1 |    98.3 |
| leather    |       99.1 |    46.4 |    46.2 |    74.1 |       99.4 |    99.5 |    99.8 |
| metal_nut  |       78.5 |    36.6 |    30.1 |    60.6 |       69.2 |    89.4 |    92.4 |
| pill       |       85   |    26.2 |    23.4 |    72.4 |       78.3 |    91.6 |    95.8 |
| screw      |       98   |    39.7 |    33.3 |    78.6 |       77   |    85.6 |    91.7 |
| tile       |       82.6 |    47.8 |    44.4 |    38.2 |       95.9 |    93.8 |    98.5 |
| toothbrush |       93.9 |    49   |    37.1 |    75.7 |       97.8 |    95.1 |    99.2 |
| transistor |       62   |    18.4 |    10.3 |    27.6 |       70.5 |    68.1 |    60.8 |
| wood       |       95.6 |    59.8 |    58.9 |    52.1 |       94.3 |    97.6 |    97.7 |
| zipper     |       85.9 |    19.6 |    16.6 |    21.6 |       77.1 |    88.1 |    93.7 |
| mean       |       86.7 |    39.8 |    35   |    54.3 |       86   |    90   |    93.3 |

I hope it helps :)

**Comment by Linda6521** [comment]
Thank you for your patience. I got it. I found that when using meta_wo_md.json to train the VisA dataset, I can get the similar results to yous. Previously, I used the meta.json to train the VisA dataset, which was universally used in other papers.

**Comment by hk-zh** [comment]
Hi, 
As we mentioned in the [issue]( https://github.com/boschresearch/MultiADS/issues/12), we actually used less data in VisA to train our model. The JSON file meta_wo_md.json means that it excludes the image with multiple types of defects in the VisA dataset. The reason for that is that we didn't label the masks for different defect types in the VisA before the paper submission. If you want to use the whole VisA dataset to train MultiADS, we actually provided the [visa extended](https://huggingface.co/datasets/zhKingg/VisA_Extended) dataset that contains masks for each defect type on one image.  Unfortunately, we have not trained on this [visa extended](https://huggingface.co/datasets/zhKingg/VisA_Extended). We will update the code base as long as we have time to implement.
